src/lib.rs

```diff
@@ -795,5 +886,252 @@ where F: FnOnce() -> R,
 
 #[test]
 fn simple_build() {
-    Builder::new().build().expect("Couldn't build default Counter");
+    Builder::new()
+        .counter()
+        .expect("Couldn't build default Counter");
+}
+
+// Use a pretty big buffer because we don't want to drop any entries
+const SAMPLE_BUFFER_PAGES: usize = 128 + 1;
+
+/// Size of a page of memory in bytes.
+fn page_size() -> usize {
+    (unsafe { libc::sysconf(libc::_SC_PAGESIZE) }) as usize
+}
+
+fn wait_for_readable_or_timeout(file: &File, timeout: Option<std::time::Duration>) -> bool {
+    let mut pollfd = pollfd {
+        fd: file.as_raw_fd(),
+        events: POLLIN,
+        revents: 0,
+    };
+    let timeout = timeout.map(|d| d.as_millis() as c_int).unwrap_or(-1);
+    let events = unsafe { poll(&mut pollfd, 1, timeout) };
+    events == 0
+}
+
+// We create our own version of this rather than use bindgen's so we can have the atomics.
+#[repr(C)]
+struct PerfEventMmapPage {
+    /// version number of this structure
+    version: u32,
+
+    /// lowest version this is compat with
+    compat_version: u32,
+
+    /// seqlock for synchronization
+    lock: u32,
+
+    /// hardware counter identifier
+    index: u32,
+
+    /// add to hardware counter value
+    offset: i64,
+
+    /// time event active
+    time_enabled: u64,
+
+    /// time event on CPU
+    time_running: u64,
+
+    capabilites: u64,
+    pmc_width: u16,
+    time_shift: u16,
+    time_mult: u32,
+    time_offset: u64,
+    __reserved: [u64; 120], /* Pad to 1 k */
+
+    /// head in the data section
+    data_head: AtomicU64,
+
+    /// user-space written tail
+    data_tail: AtomicU64,
+
+    /// where the buffer starts
+    data_offset: u64,
+
+    /// data buffer size
+    data_size: u64,
+
+    aux_head: u64,
+    aux_tail: u64,
+    aux_offset: u64,
+    aux_size: u64,
+}
+
+/// A stream of samples being sent to us from the kernel. These samples represent instantaneous
+/// states or events concerning the process(es) being profiled.
+///
+/// Internally the samples are queued up in a ring-buffer. The kernel writes samples into the
+/// buffer, and [`SampleStream.read`] dequeues them. If the buffer is full, the kernel will
+/// overwrite old samples effectively dropping them.
+pub struct SampleStream {
+    file: File,
+    mapped_memory: *mut c_void,
+    attrs: sys::bindings::perf_event_attr,
+}
+
+unsafe impl Send for SampleStream {}
+unsafe impl Sync for SampleStream {}
+
+impl SampleStream {
+    fn new(attrs: sys::bindings::perf_event_attr, file: File) -> std::io::Result<Self> {
+        let mapped_memory = check_syscall(|| unsafe {
+            mmap(
+                std::ptr::null_mut(),
+                SAMPLE_BUFFER_PAGES * page_size(),
+                PROT_READ | PROT_WRITE,
+                MAP_SHARED,
+                file.as_raw_fd(),
+                0,
+            ) as isize
```

Comment:
I've personally used the `memmap` crate for convenience, I wonder how much sense that would make here?